vendor/zkllvm-metacraft-circuits/src/circuit_utils/circuit_byte_utils.h

```diff
@@ -0,0 +1,187 @@
+#pragma once
+
+#include <nil/crypto3/hash/algorithm/hash.hpp>
+#include <nil/crypto3/hash/sha2.hpp>
+#include "base_types.h"
+#include "../utils/picosha2.h"
+
+using namespace nil::crypto3;
+
+namespace circuit_byte_utils {
+
+#define countof(array) (sizeof(array) / sizeof(array[0]))
+
+    using sha256_t = typename hashes::sha2<256>::block_type;
+
+    bool sha256_equals(sha256_t hash1, sha256_t hash2) {
+
+        bool result = true;
+        for (auto i = 0; i < countof(hash1); ++i) {
+            result = result && (hash1[i] == hash2[i]);
+        }
+
+        return result;
+    }
+
+    template<typename T>
+    Byte get_nth_byte(const T val, unsigned int n) {
+        static_assert(std::is_integral<typename std::remove_reference<T>::type>::value, "T must be integral");
+        assert_true(n < sizeof(T));
+
+        return T(val >> (n * 8));
+    }
+
+    bool get_nth_bit(uint64_t value, short i) {
+        return bool(1 & (value >> i));
+    }
+
+    Byte set_nth_bit(Byte& value, short i) {
+        assert_true(i < 8);
+        value = value | (Byte(1) << i);
+        return value;
+    }
+
+    template<typename T>
+    void sha256_to_bytes_array(sha256_t sha, T& out) {
```

Comment:
Make this function non-template. Rename this to sha256_to_bytes. Make this return a Bytes32 instead of accepting it as an out parameter.